stores/sql/sqlite/main.go

```diff
@@ -98,7 +108,200 @@ func (tx *MainDatabaseTx) DeleteObjects(ctx context.Context, bucket string, key
+		}
+	}
+
+	// insert slices
+	insertSliceStmt, err := tx.Prepare(ctx, `INSERT INTO slices (created_at, db_object_id, object_index, db_multipart_part_id, db_slab_id, offset, length)
+								VALUES (?, ?, ?, ?, ?, ?, ?)`)
+	if err != nil {
+		return fmt.Errorf("failed to prepare statement to insert slice: %w", err)
+	}
+	defer insertSliceStmt.Close()
+
+	for i := range slices {
+		res, err := insertSliceStmt.Exec(ctx,
+			time.Now(),
+			objID,
+			uint(i+1),
+			nil,
+			slabIDs[i],
+			slices[i].Offset,
+			slices[i].Length,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to insert slice: %w", err)
+		} else if n, err := res.RowsAffected(); err != nil {
+			return fmt.Errorf("failed to get rows affected: %w", err)
+		} else if n == 0 {
+			return fmt.Errorf("failed to insert slice: no rows affected")
+		}
+	}
+
+	// insert sectors
+	insertSectorStmt, err := tx.Prepare(ctx, `INSERT INTO sectors (created_at, db_slab_id, slab_index, latest_host, root)
+								VALUES (?, ?, ?, ?, ?) ON CONFLICT(root) DO UPDATE SET latest_host = EXCLUDED.latest_host RETURNING id, db_slab_id`)
+	if err != nil {
+		return fmt.Errorf("failed to prepare statement to insert sector: %w", err)
+	}
+	defer insertSectorStmt.Close()
+
+	var sectorIDs []int64
+	for i, ss := range slices {
+		for j := range ss.Shards {
+			var sectorID, slabID int64
+			err := insertSectorStmt.QueryRow(ctx,
+				time.Now(),
+				slabIDs[i],
+				j+1,
+				ssql.PublicKey(ss.Shards[j].LatestHost),
+				ss.Shards[j].Root[:],
+			).Scan(&sectorID, &slabID)
+			if err != nil {
+				return fmt.Errorf("failed to insert sector: %w", err)
+			} else if slabID != slabIDs[i] {
+				return fmt.Errorf("failed to insert sector for slab %v: already exists for slab %v", slabIDs[i], slabID)
+			}
+			sectorIDs = append(sectorIDs, sectorID)
+		}
+	}
+
+	// insert contract <-> sector links
+	insertContractSectorStmt, err := tx.Prepare(ctx, `INSERT INTO contract_sectors (db_sector_id, db_contract_id)
+											VALUES (?, ?) ON CONFLICT(db_sector_id, db_contract_id) DO NOTHING`)
+	if err != nil {
+		return fmt.Errorf("failed to prepare statement to insert contract sector link: %w", err)
+	}
+	defer insertContractSectorStmt.Close()
+
+	sectorIdx := 0
+	for _, ss := range slices {
+		for _, shard := range ss.Shards {
+			for _, fcids := range shard.Contracts {
+				for _, fcid := range fcids {
+					if _, ok := usedContracts[fcid]; ok {
+						_, err := insertContractSectorStmt.Exec(ctx,
+							sectorIDs[sectorIdx],
+							usedContracts[fcid].ID,
+						)
+						if err != nil {
+							return fmt.Errorf("failed to insert contract sector link: %w", err)
+						}
+					} else {
+						tx.log.Warn("missing contract for shard",
+							"contract", fcid,
+							"root", shard.Root,
+							"latest_host", shard.LatestHost,
+						)
+					}
+				}
+			}
+			sectorIdx++
+		}
+	}
+
+	// update metadata
+	if _, err := tx.Exec(ctx, "DELETE FROM object_user_metadata WHERE db_object_id = ?", objID); err != nil {
```

Comment:
I take it we can't `UPSERT`? Kind of sad to do the delete and re-insert.

Reply:
I think we can actually just get rid of the delete. `InsertObject` would fail if an object already exists and since the object doesn't exist, the metadata doesn't exist either due to the foreign key constraint.